### PR TITLE
[MPS] GridSampler2D backward

### PR DESCRIPTION
**Pull request by malfet**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #179756

Implements `grid_sampler_2d_backward_mps` as Metal kernel, ported from the CUDA implementation in GridSampler.cu / GridSampler.cuh. Supports all combinations of interpolation and padding modes  for float32, float16, and bfloat16. Has been requested more than 10 times in https://github.com/pytorch/pytorch/issues/141287

## Benchmark: `grid_sampler_2d` backward on Apple M5 Pro

<details>
<summary>Benchmark script</summary>

```python
"""Benchmark for grid_sampler_2d backward on MPS."""
import time, torch

def bench(N, C, H, W, interp=0, dtype=torch.float32, memory_format=torch.contiguous_format,
          warmup=10, repeat=100):
    inp = torch.randn(N, C, H, W, device="mps", dtype=dtype)
    inp = inp.to(memory_format=memory_format).requires_grad_(True)
    grid = torch.randn(N, H, W, 2, device="mps", dtype=dtype)
    out = torch.grid_sampler_2d(inp, grid, interp, 0, True)
    grad = torch.randn_like(out)
    args = (grad, inp, grid, interp, 0, True, [True, True])
    for _ in range(warmup):
        torch.ops.aten.grid_sampler_2d_backward(*args)
    torch.mps.synchronize()
    times = []
    for _ in range(repeat):
        torch.mps.synchronize()
        t0 = time.perf_counter()
        torch.ops.aten.grid_sampler_2d_backward(*args)
        torch.mps.synchronize()
        times.append(time.perf_counter() - t0)
    times.sort()
    t = max(1, len(times) // 10)
    trimmed = times[t:-t]
    return sum(trimmed) / len(trimmed) * 1000

SHAPES = [(1,32,64,64), (1,16,256,256), (8,64,128,128), (16,128,256,256)]
INTERPS = {0: "bilinear", 2: "bicubic"}
FORMATS = {"contiguous": torch.contiguous_format, "channels_last": torch.channels_last}

print(f"{'dtype':<10} {'interp':<10} {'layout':<14} {'(N,C,H,W)':<24} {'ms':>8}")
print("-" * 70)
for dtype in [torch.float32, torch.float16]:
    for interp, iname in INTERPS.items():
        for fname, fmt in FORMATS.items():
            for shape in SHAPES:
                avg = bench(*shape, interp=interp, dtype=dtype, memory_format=fmt)
                print(f"{str(dtype):<10} {iname:<10} {fname:<14} {str(shape):<24} {avg:8.3f}")
```

</details>

| dtype | interp | (N,C,H,W) | contiguous (ms) | channels_last (ms) |
|-------|--------|-----------|----:|----:|
| float32 | bilinear | (1, 32, 64, 64) | 0.285 | 0.249 |
| float32 | bilinear | (1, 16, 256, 256) | 0.351 | 0.332 |
| float32 | bilinear | (8, 64, 128, 128) | 1.809 | 1.717 |
| float32 | bilinear | (16, 128, 256, 256) | 81.720 | 62.959 |
| float32 | bicubic | (1, 32, 64, 64) | 0.434 | 0.517 |
| float32 | bicubic | (1, 16, 256, 256) | 0.819 | 0.952 |
| float32 | bicubic | (8, 64, 128, 128) | 4.527 | 5.331 |
| float32 | bicubic | (16, 128, 256, 256) | 168.363 | 145.735 |
| float16 | bilinear | (1, 32, 64, 64) | 0.291 | 0.295 |
| float16 | bilinear | (1, 16, 256, 256) | 0.462 | 0.431 |
| float16 | bilinear | (8, 64, 128, 128) | 2.080 | 2.052 |
| float16 | bilinear | (16, 128, 256, 256) | 31.970 | 27.160 |
| float16 | bicubic | (1, 32, 64, 64) | 0.774 | 0.769 |
| float16 | bicubic | (1, 16, 256, 256) | 1.219 | 1.282 |
| float16 | bicubic | (8, 64, 128, 128) | 7.032 | 8.497 |
| float16 | bicubic | (16, 128, 256, 256) | 104.402 | 170.552 |

**Notes:**
- Bilinear channels_last is up to 23% faster than contiguous (62.9 vs 81.7ms for the largest float32 case)
- Bicubic channels_last regresses for float16 large tensors (170ms vs 104ms) due to the 4x4 spatial access pattern losing locality

Co-Authored-By: Claude Opus 4.6 <noreply@anthropic.com>

### CODE REVIEW

File: aten/src/ATen/native/mps/kernels/GridSampler.metal [hunk 1]
```diff
@@ -608,3 +624,488 @@ kernel void grid_sampler_3d(
...
+  }
+
+  auto gGrid_ptr_NHW = grad_grid + tid * gGrid_sW;
+  gGrid_ptr_NHW[0] = static_cast<T>(ix.y * gix);
+  gGrid_ptr_NHW[1] = static_cast<T>(iy.y * giy);
+}
+
+// Registration macros for backward kernels
+#define REGISTER_GRID_SAMPLER_2D_BACKWARD(DTYPE, INTERP, INAME, PAD, PNAME) \
+  template [[host_name("grid_sampler_2d_backward_" INAME "_" PNAME          \
```
Comment (kurtamohler) [inline aten/src/ATen/native/mps/kernels/GridSampler.metal]: `INTERP` and `INAME` seem redundant. It should be possible to do this, right?


```suggestion
#define REGISTER_GRID_SAMPLER_2D_BACKWARD(DTYPE, INTERP, PAD, PNAME) \
  template [[host_name("grid_sampler_2d_backward_" #INTERP "_" PNAME          \
```

Comment (malfet) [reply aten/src/ATen/native/mps/kernels/GridSampler.metal]: Good point, let me do this one (and few other cleanups, for example makes no sense to init empty tensor strides, and I wonder if I can just use nullptr to indicate that input tensor is not used)

File: aten/src/ATen/native/mps/kernels/GridSampler.metal [hunk 2]
```diff
@@ -608,3 +624,488 @@ kernel void grid_sampler_3d(
...
+  coeffs[1] = (-3 * (A + 2) * x - 2 * (A + 3)) * x;
+  x = 1 - t;
+  coeffs[2] = (3 * (A + 2) * x - 2 * (A + 3)) * x;
+  x = 2 - t;
+  coeffs[3] = (3 * A * x - 10 * A) * x + 8 * A;
+}
+
+// Bilinear backward kernel
+template <typename Pad, typename T>
+kernel void grid_sampler_2d_backward_bilinear(
```
Comment (kurtamohler) [inline aten/src/ATen/native/mps/kernels/GridSampler.metal]: It would be nice to combine the bilinear, bicubic, and nearest kernels into one, with a function pointer template parameter to switch between them, since there are a lot of identical lines of code in the three kernels. But maybe it's not worth the trouble, if the function pointer would need a huge list of arguments

Comment (malfet) [reply aten/src/ATen/native/mps/kernels/GridSampler.metal]: Have you tested which one is more performant? I know you pass mode as an argument to grid_sampler_3d, but for some reason CUDA follows this pattern for both interpolate and 2d... Template argument do sound reasonable to me

**Review by manuelcandales — APPROVED** [review]

File: aten/src/ATen/native/mps/kernels/GridSampler.metal (hunk 2 above)
Comment (malfet) [reply aten/src/ATen/native/mps/kernels/GridSampler.metal]: I can benchmark it as followup PR. This one is a literal transpiling of CUDA shader into Metal without significant architectural changes. But if selecting interpolation mode dynamically and statically (via kernel dispatch) have the same perf, it indeed stands to reason to use it

**Comment by mlaves** [comment]
> > I suppose it's worth mentioning that #179388 was submitted to implement this already. But I'm guessing that this one might give better performance since it can calculate both grads in the same kernel
> 
> I waited for the author to fix numeric for some time (even in their original #159421 issue), but it never happened

Hi, thanks for opening this PR. My original PR #159421 did not have any numerical issues and all tests were passing. I removed the GridSampler2D backward from my current PR #179388 to avoid conflicts with this PR.

File: aten/src/ATen/native/mps/kernels/GridSampler.metal (hunk 2 above)
Comment (kurtamohler) [reply aten/src/ATen/native/mps/kernels/GridSampler.metal]: I haven't tested the performance. I had assumed that the compiler would make them the same, but I agree that there is a chance that it doesn't

**Review by kurtamohler — APPROVED** [review]

File: aten/src/ATen/native/mps/operations/GridSampler.mm [hunk 3]
```diff
@@ -223,4 +226,97 @@ Tensor grid_sampler_3d_mps(const Tensor& input,
...
+      } else {
+        mtl_setArgs(computeEncoder, input, grad_grid, grad_output, input, grid, params);
+      }
+
+      mtl_dispatch1DJob(computeEncoder, pso, num_threads);
+      getMPSProfiler().endProfileKernel(pso);
+    }
+  });
+
+  return std::make_tuple(grad_input, grad_grid);
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/operations/GridSampler.mm]:
```suggestion
  return std::make_tuple(std::move(grad_input), std::move(grad_grid));
